### PR TITLE
Azure-pipelines integration

### PR DESCRIPTION
**Pull request by thijsmie**
This commit adds the pipelines to test, lint, publish and report coverage. Building and publishing documentation will follow as soon as we have set it up on the C side.
To merge this a few extra external steps are required:

- Set up the repo in azure-pipelines
- Add a CODECOV_TOKEN environment variable
- Add a pypi_credentials secure file for publishing
- Add a PYPI_REPOSITORY environment variable for publishing

The last two can wait, the pipeline won't run until we tag a release.

For a full example of the pipeline runs see: https://dev.azure.com/tmiedema/cyclonedds-python/_build/results?buildId=109&view=results

N.B. There are a few actual C code changes in this pull request as well. These fix a bug introduced by the segfault fix that only occurs on windows by what I can only assume is wonky compiler behaviour.

### CODE REVIEW

File: .azure/publish-coverage.yml
Comment (k0ekk0ek) [inline .azure/publish-coverage.yml:47]: https://about.codecov.io/security-update/

Comment (thijsmie) [reply .azure/publish-coverage.yml:47]: I will update this with the curl-uploader as used in the main cyclone repo

Comment (thijsmie) [reply .azure/publish-coverage.yml]: Solved by using `codecov-python`, sadly depending on my own fork of it as the Azure Pipelines configuration was not present. I will submit a pull request and once that is merged and released we can change over to the proper package.
See here the report from a succesful run: https://codecov.io/gh/thijsmie/cyclonedds-python/commit/c2fdeafebbd5a098dc046e5ccbe811637a2887ce/

**Review by eboasson — APPROVED** [review]
LGTM, thanks!

@k0ekk0ek perhaps you want to have a final look and merge it if you're now ok with it, too?

File: src/cyclonedds/clayer/src/pysertype.c
Comment (k0ekk0ek) [inline src/cyclonedds/clayer/src/pysertype.c:1524]: There's quite a lot of repetition here. I'm fine with it for now, but at some point it might be worth going about this in a more practical fashion. At least, seems like this sort of thing happens all over the code.

File: .azure/publish-coverage.yml
Comment (thijsmie) [reply .azure/publish-coverage.yml:1]: I avoided this name because it actually has a special meaning for codecov as a [settings file](https://docs.codecov.io/docs/codecov-yaml) so if it gets uploaded to codecov itself by accident the codecov analysis will fail. If that risk is not important to you I will change it.

File: .azure/publish-release.yml
Comment (thijsmie) [reply .azure/publish-release.yml:20]: This comes from a copy from a personal project, where I use tags for other things as just versions, which get explicitly prefixed with a "v". I will change it.

File: src/cyclonedds/clayer/src/pysertype.c
Comment (thijsmie) [reply src/cyclonedds/clayer/src/pysertype.c:1524]: I agree the error handling like this is suboptimal, we should definitely pursue a better style for this, a thing for the todo list.

File: .azure/publish-coverage.yml
Comment (k0ekk0ek) [reply .azure/publish-coverage.yml:1]: Given that it's in the `.azure` directory, changes of that happening seem non-existent.

Comment (k0ekk0ek) [reply .azure/publish-coverage.yml:1]: It'll do for now, let's get CI setup first...

File: .azure/publish-coverage.yml
Comment (k0ekk0ek) [inline .azure/publish-coverage.yml:42]: I do have a bit of a problem with the fact this references your private fork (not that Codecov has the best track-record when it comes to security). At the very least I'd like to see a ticket with an attempt being made to get these changes upstream.

Comment (thijsmie) [reply .azure/publish-coverage.yml:42]: I also have a problem with it being a private fork and I have submitted a PR a while ago, see: [codecov-python!319](https://github.com/codecov/codecov-python/pull/319). Sadly there has been no response from the maintainers for a month now, I just pinged them again.

Do you want me to drop the python uploader and switch back to the bash version we use for the main repo?

Comment (k0ekk0ek) [reply .azure/publish-coverage.yml:42]: It's up to you. I'm fine with the link just being there, we can always switch back. But fyi, we don't use the bash uploader for either *cyclonedds* or *cyclonedds-cxx*. [I had a feeling just executing a downloaded script might not be the best idea](https://github.com/eclipse-cyclonedds/cyclonedds/pull/184#issuecomment-493377004).

**Pull request merged**
